### ISSUE TITLE
Intel MKL FATAL ERROR in 20.11 and 20.12

### ISSUE THREAD

**Issue by HarmitMinhas96** [triton-inference-server/server#2379]
**Description**
- When trying to run a PyTorch script which was traced on CPU, the error: `Intel MKL FATAL ERROR: Cannot load libmkl_avx2.so or libmkl_def.so` is thrown. This is a CPU script that is working on 20.10, but fails on 20.11 and 20.12. After looking around a bit, I was able to get around that error by setting `LD_PRELOAD="/opt/tritonserver/backends/pytorch/libmkl_avx2.so:/opt/tritonserver/backends/pytorch/libmkl_def.so"`, but now a different but similar error is thrown:
`Intel MKL FATAL ERROR: cannot load libmkl_vml_avx2.so or libmkl_vml_def.so.`.

When looking in the 20.12 container, I get nothing:
```
root@py-svc-triton-server-669bf765b-svvzm:/opt/tritonserver# find . -name '*libmkl_vml*'
root@py-svc-triton-server-669bf765b-svvzm:/opt/tritonserver# 
```
When looking in the 20.10 container, I can find one of the files from that error (which seems to be enough):
```
root@py-svc-triton-server-f744f4657-mqmk5:/opt/tritonserver# find . -name '*libmkl_vml*'
./lib/pytorch/libmkl_vml_def.so
root@py-svc-triton-server-f744f4657-mqmk5:/opt/tritonserver# 
```

**Triton Information**
What version of Triton are you using?
- Currently using 20.10 and trying to upgrade to 20.12
Are you using the Triton container or did you build it yourself?
- Using the triton container

**To Reproduce**
Steps to reproduce the behavior.

Describe the models (framework, inputs, outputs), ideally include the model configuration file (if using an ensemble include the model configuration file for that as well).
- Using a cpu traced PyTorch script using torch==1.5. 

**Expected behavior**
- Expect a cpu trace PyTorch script which ran on 20.10 to continue working on 20.11 and 20.12.

**Comment by CoderHam** [comment]
You can run the model on CPU even after tracing on GPU by specifying the device type in the model config:
```
instance_group [
  {
    kind: KIND_CPU
  }
]
```

Could you try tracing the model on GPU and then running the same on CPU as above?

**Comment by HarmitMinhas96** [comment]
I ran the GPU trace with the CPU instance group set, in a container that doesn't have access to GPU's and got:
`RuntimeError: CUDA error: CUDA driver version is insufficient for CUDA runtime version`.

**Comment by CoderHam** [comment]
The container will need access to the GPU to trace the model on GPU. There is no workaround for this.

**Comment by HarmitMinhas96** [comment]
The model was traced on GPU and works fine in a container that has access to a GPU. As per your suggestion to workaround my CPU trace issue, I tried to run the GPU traced model on CPU, but that does not work.

**Comment by HarmitMinhas96** [comment]
Also, as per pytorch docs, this isn't ideal: https://pytorch.org/docs/stable/jit.html#frequently-asked-questions

**Comment by Debangshu132** [comment]
Same issue here, does not work in 20.11 or 20.12 but works fine in 20.10

**Comment by CoderHam** [comment]
I repeated the experiment for resnet50 using both torch.jit.script and torch.jit.trace on a CPU only system and the model.pt produced works fine in tritonserver in both 20.11 and 20.12. This is the code I used to produce both the model files.
```
import torch
import torchvision

model = torchvision.models.resnet50(pretrained=True)
model.eval()

example_input = torch.rand(1, 3, 224, 224, dtype=torch.float)
traced = torch.jit.trace(model, example_input)
traced.save("traced_model.pt")

scripted = torch.jit.script(model)
scripted.save("scripted_model.pt")

```
@Debangshu132 and @HarmitMinhas96 can you share the model you are tracing when you see this error. 
Ps: I used the PyTorch 20.11 NGC container to produce the traced/scripted models.

**Comment by adamm123** [comment]
I have a similar problem, but I try to evaluate a model on the GPU. FasterRCNN model works on 20.10 but fails on 20.11 and 20.12.

I evaluated the model from the comment above and it works fine on all three versions of Triton.

> I repeated the experiment for resnet50 using both `torch.jit.script` and `torch.jit.trace` on a CPU only system and the model.pt produced works fine in tritonserver (both 20.11 and 20.12). This is the code I used to produce both the model files.
> 
> ```
> import torch
> import torchvision
> 
> model = torchvision.models.resnet50(pretrained=True)
> model.eval()
> 
> example_input = torch.rand(1, 3, 224, 224, dtype=torch.float)
> traced = torch.jit.trace(model, example_input)
> traced.save("traced_model.pt")
> 
> scripted = torch.jit.script(model)
> scripted.save("scripted_model.pt")
> ```
> 
> @Debangshu132 and @HarmitMinhas96 can you share the model you are tracing when you see this error.
> Ps: I used the PyTorch 20.11 NGC container to produce the traced/scripted models.

However,  when I try to run FasterRCNN model it fails.
model export:
```python
import torch
import torchvision


class FasterRCNNWrapper(torch.nn.Module):

    def __init__(self):
        super().__init__()

        self.model = torchvision.models.detection.fasterrcnn_resnet50_fpn(pretrained=True)
        self.model = self.model.eval()

    def forward(self, x):
        losses, detections = self.model([x])
        return detections[0]['boxes'], detections[0]['labels'], detections[0]['scores']


model = FasterRCNNWrapper()

script = torch.jit.script(model)
script.save('model.pt')

```
config.pbtxt:
```
name: "od"
platform: "pytorch_libtorch"
input [
  {
    name: "inputs__0"
    data_type: TYPE_FP32
    dims: [ 3, -1, -1 ]
  }
]
output [
  {
    name: "boxes__0"
    data_type: TYPE_FP32
    dims: [ 1000, 4  ]
  },
  {
    name: "labels__1"
    data_type: TYPE_INT64
    dims: [ 1000 ]
  },
  {
    name: "scores__2"
    data_type: TYPE_FP32
    dims: [ 1000 ]
  }
]
instance_group [
  {
    kind: KIND_GPU
  }
]

```

stacktrace:
```
I0112 07:59:47.980297 1 http_server.cc:1229] HTTP request: 2 /v2/models/od/versions/1/infer
I0112 07:59:47.980375 1 model_repository_manager.cc:496] GetInferenceBackend() 'od' version 1
I0112 07:59:47.980401 1 model_repository_manager.cc:496] GetInferenceBackend() 'od' version 1
I0112 07:59:47.980496 1 infer_request.cc:502] prepared: [0x0x7f62bc00a9c0] request id: , model: od, requested version: 1, actual version: 1, flags: 0x0, correlation id: 0, batch size: 0, priority: 0, timeout (us): 0
original inputs:
[0x0x7f62bc0083a8] input: inputs__0, type: FP32, original shape: [3,1467,2000], batch + shape: [3,1467,2000], shape: [3,1467,2000]
override inputs:
inputs:
[0x0x7f62bc0083a8] input: inputs__0, type: FP32, original shape: [3,1467,2000], batch + shape: [3,1467,2000], shape: [3,1467,2000]
original requested outputs:
boxes__0
labels__1
scores__2
requested outputs:
boxes__0
labels__1
scores__2

I0112 07:59:47.981069 1 libtorch.cc:1100] model od, instance od_0, executing 1 requests
I0112 07:59:47.981115 1 libtorch.cc:510] TRITONBACKEND_ModelExecute: Running od_0 with 1 requests
INTEL MKL ERROR: /opt/tritonserver/backends/pytorch/libmkl_vml_avx2.so: cannot open shared object file: No such file or directory.
Intel MKL FATAL ERROR: cannot load libmkl_vml_avx2.so or libmkl_vml_def.so.
+0112 07:59:50.112158 1 grpc_server.cc:270] Process for CudaSharedMemoryUnregister, rpc_ok=0, 0 step START
+0112 07:59:50.112211 1 grpc_server.cc:408] Done for CudaSharedMemoryUnregister, 0
+0112 07:59:50.112249 1 grpc_server.cc:270] Process for CudaSharedMemoryRegister, rpc_ok=0, 0 step START
+0112 07:59:50.112265 1 grpc_server.cc:408] Done for CudaSharedMemoryRegister, 0
+0112 07:59:50.112288 1 grpc_server.cc:270] Process for CudaSharedMemoryStatus, rpc_ok=0, 0 step START
+0112 07:59:50.112305 1 grpc_server.cc:408] Done for CudaSharedMemoryStatus, 0
+0112 07:59:50.112332 1 grpc_server.cc:270] Process for SystemSharedMemoryUnregister, rpc_ok=0, 0 step START
+0112 07:59:50.112348 1 grpc_server.cc:408] Done for SystemSharedMemoryUnregister, 0
+0112 07:59:50.112371 1 grpc_server.cc:270] Process for SystemSharedMemoryRegister, rpc_ok=0, 0 step START
+0112 07:59:50.112387 1 grpc_server.cc:408] Done for SystemSharedMemoryRegister, 0
+0112 07:59:50.112409 1 grpc_server.cc:270] Process for SystemSharedMemoryStatus, rpc_ok=0, 0 step START
+0112 07:59:50.112426 1 grpc_server.cc:408] Done for SystemSharedMemoryStatus, 0
+0112 07:59:50.112461 1 grpc_server.cc:270] Process for RepositoryModelUnload, rpc_ok=0, 0 step START
+0112 07:59:50.112478 1 grpc_server.cc:408] Done for RepositoryModelUnload, 0
+0112 07:59:50.112497 1 grpc_server.cc:270] Process for RepositoryModelLoad, rpc_ok=0, 0 step START
+0112 07:59:50.112513 1 grpc_server.cc:408] Done for RepositoryModelLoad, 0
+0112 07:59:50.112531 1 grpc_server.cc:270] Process for RepositoryIndex, rpc_ok=0, 0 step START
+0112 07:59:50.112547 1 grpc_server.cc:408] Done for RepositoryIndex, 0
+0112 07:59:50.112567 1 grpc_server.cc:270] Process for ModelStatistics, rpc_ok=0, 0 step START
+0112 07:59:50.112582 1 grpc_server.cc:408] Done for ModelStatistics, 0
+0112 07:59:50.112675 1 grpc_server.cc:270] Process for ModelConfig, rpc_ok=0, 0 step START
+0112 07:59:50.112708 1 grpc_server.cc:408] Done for ModelConfig, 0
+0112 07:59:50.112744 1 grpc_server.cc:270] Process for ModelMetadata, rpc_ok=0, 0 step START
+0112 07:59:50.112760 1 grpc_server.cc:408] Done for ModelMetadata, 0
+0112 07:59:50.112750 1 grpc_server.cc:3434] Process for ModelStreamInferHandler, rpc_ok=0, context 2, 3 step START
+0112 07:59:50.112779 1 grpc_server.cc:270] Process for ServerMetadata, rpc_ok=0, 0 step START+0112 07:59:50.112787 1 grpc_server.cc:3089] Process for ModelInferHandler, rpc_ok=0, 1 step START+0112 07:59:50.112797 1 grpc_server.cc:2139] Done for ModelStreamInferHandler, 3


+0112 07:59:50.112831 1 grpc_server.cc:2139] Done for ModelInferHandler, 1
+0112 07:59:50.112845 1 grpc_server.cc:408] Done for ServerMetadata, 0
+0112 07:59:50.112865 1 grpc_server.cc:270] Process for ModelReady, rpc_ok=0, 0 step START
+0112 07:59:50.112881 1 grpc_server.cc:408] Done for ModelReady, 0
+0112 07:59:50.112900 1 grpc_server.cc:270] Process for ServerReady, rpc_ok=0, 0 step START
+0112 07:59:50.112915 1 grpc_server.cc:408] Done for ServerReady, 0
+0112 07:59:50.112943 1 grpc_server.cc:270] Process for ServerLive, rpc_ok=0, 0 step START
+0112 07:59:50.112960 1 grpc_server.cc:408] Done for ServerLive, 0
+0112 07:59:50.113116 1 grpc_server.cc:426] Thread exited for CommonHandler
+0112 07:59:50.113172 1 grpc_server.cc:2160] Thread exited for ModelInferHandler
+0112 07:59:50.113216 1 grpc_server.cc:2160] Thread exited for ModelStreamInferHandler
+0112 07:59:50.113256 1 grpc_server.cc:2115] Destructed ModelStreamInferHandler
+0112 07:59:50.113275 1 grpc_server.cc:2115] Destructed ModelInferHandler
```
Triton does not recover from this error, just freezes. On client side request ends with timeout.
Sending another request to Triton results in `ConnectionRefusedError: [Errno 111] Connection refused`.

**Comment by CoderHam** [comment]
@adamm123 thanks for getting back to me. I will test with the FasterRCNN model and have a fix up shortly.

**Comment by CoderHam** [comment]
A few issues: Tracing RCNN models is the better approach than scripting. Additionally the FasterRCNNWrapper model needs to be set to eval() as well. 
https://github.com/pytorch/vision/issues/1002#issuecomment-639370764

I was able to reproduce the issue. The fix is coming up in #2406 
But you will need to fix the model generation script otherwise you will see an error originating from here: https://github.com/pytorch/vision/blob/11a39aaab5b55a3c116c2e8d8001bad94a96f99d/torchvision/models/detection/generalized_rcnn.py#L108
`[W faster_rcnn.py:95] Warning: RCNN always returns a (Losses, Detections) tuple in scripting (function )`

**Comment by HarmitMinhas96** [comment]
@CoderHam Thank you for the fix!
Sorry I was not able to get back with a reproduce-able example
